### DISCUSSION TITLE
Voice cloning

### DISCUSSION THREAD

**Discussion by astricks** [coqui-ai/TTS#653] [General Q&A]
Hi,

First off, my apologies for bringing up voice cloning yet again since I see a few threads and issues on it already.

My reasons for starting a discussion are
1. I'd like to try voice-cloning and then document the steps for it so that it avoids another issue/discussion.
2. I have questions regarding how voice cloning works, specifically as it relates to languages. I'd like to try cloning a voice in the Hindi language. If successful, I can add the example to the proposed documentation.
3. There now seems to be a [substantially better speaker encoder](https://github.com/coqui-ai/TTS/pull/508) thanks to @Edresson which might make voice cloning much more accurate

My questions are as follows:

1. For very accurate voice cloning, I understand that all 3 components (speaker_encoder, TTS model & vocoder) need to be trained on (ideally non-overlapping) datasets containing 1000s of speakers.

   **For Coqui, what are the models that can be recommended for decent cloning results?**

I assume for speaker encoder, the model @Edresson is training would be best. For vocoder, we have universal vocoders which work well. What about TTS?

2. On the topic of TTS - would the TTS model need to be trained on 1000s of speakers in the language of interest for it to convincingly clone in that language?

3. If so, multispeaker datasets in non-English languages are rare - is there a technique to overcome this limitation?

4. @Edresson very excited to try the new speaker encoder, would you be sharing it soon?

Thanks in advance, and hope to learn from this discussion and contribute back.

**Comment by loganhart02** [comment]
to understand voice cloning as whole better look at this repo if you haven;t already https://github.com/CorentinJ/Real-Time-Voice-Cloning
working on the same thing but for english language so I don't have the data hurdles you have but this is the only implimentation from the original paper which is also a good read https://arxiv.org/pdf/1806.04558.pdf

**Reply by loganhart02** [reply]
you should also try looking at sc-glow-tts paper by @Edresson, its more promising in my opinion https://arxiv.org/pdf/2104.05557.pdf Two models you'll want to use from this repo for these papers and a voice cloner TTS model will be tacotron2 and glow-tts and make sure the configs are accurate to the respective papers

**Reply by astricks** [reply]
I did look at CorentinJ's repo/thesis, it is based on Tactron as well.

Currently, I am using Tacotron2 for TTS because it seems to generalize easily for broader use-cases and configurations. I'll try experimenting with glow-TTS at some point.

**Reply by astricks** [reply]
Just skimmed through the paper, looks very promising!

Have you been able to achieve generic voice cloning with either Tacotron or sc-glow-tts? I'd like to try to replicate it in any language. If yes, could you please share what models you used?

**Reply by astricks** [reply]
Hey @loganhart420 were you able to get good voice-cloning results using GlowTTS?

**Reply by loganhart02** [reply]
The result were okay but I ditched the model in favor of training a new one on the new speaker encoder.

**Reply by astricks** [reply]
Gotcha, thanks. If possible, would you be able to share which TTS version you used, and your config/runner file?

**Reply by loganhart02** [reply]
I used version 0.0.13( The one the model was first released on) and I just used the same config as the pretrained model and changed some of the audio params to try and fit the LibriTTS dataset(although I didn't like the results). I don't have my actual config because I usually just delete models I don't like the results of right after. I'm also adding a ScGlowTTS to the AutoTTS files so you can experiment and train it without having to write a lot of code.

**Comment by erogol** [comment]
Current SC-GlowTTS does not work perfectly with novel speakers. It requires a better speaker encoder and training the glowTTS model on a larger multi-speaker dataset like libriTTS.

@Edresson  correct if I am wrong.

**Comment by Edresson** [comment]
Hi @astricks 

The speaker encoder that I trained using all the common voice + Voxceleb datasets will be made available along with an upcoming article.

About voice cloning, if you want to clone just one voice it is recommended to just train a single speaker model in the target language, after just fine-tuning the model to the desired voice (creating a small dataset with this voice ~2 hours). With the GlowTTS model, this fine-tuning can be done with less data, since only the model's decoder learns speaker-dependent information. For this reason, in GlowTTS you can freeze the model encoder and just train the decoder to adapt to a new voice.  In this way avoiding the loss of the model vocabulary.

On the other hand, if you want generic cloning you will need thousands of speakers in the target language. In the literature, this problem hasn't been solved yet (we're working on it may be soon we'll have a solution :))

As a model for zero-shot multispeaker  TTS, I recommend [SC-GlowTTS](https://arxiv.org/abs/2104.05557) (GlowTTS with external speaker embeddings) because the results we present in the article are better than Tacotron 2, comparing in the same dataset and with the same speaker encoder.

**Reply by astricks** [reply]
Hi @Edresson,
Good to hear about the speaker encoder, looking forward it using it!

My goal is to achieve generic voice cloning, ideally inter-language. Taking baby steps, I'm thinking I'll start with voice cloning in English (because of the availability of datasets) and then try to tackle Hindi (and other Indian languages).

I skimmed through the SC-GlowTTS paper and it looks very promising! I'll give SC_glowTTS a try, see how that goes, and report back. Thanks!

**Comment by organic-chemistry** [comment]
Hello, I would like to use your work to do voice cloning .
For this I would like to first encode a vector from an external wave
And then use the tts_models/en/vctk/sc-glow-tts model.
I tried to go read the code and if I understood correctly, the defaulft encoder in your package:
TTS/speaker_encoder/configs/config.json is the one that correspond to the model I mentionned.
So to make it work I have to generate the vector corresponding to the wav sound file and then use it.
I tried a bit to modified the code in order to use it more directly like that:

`tts --text "This is a first step for humanity" --model_name  tts_models/en/vctk/sc-glow-tts --speaker_wav mp376.wav --encoder_config_path /home/jarbona/.local/lib/python3.8/site-packages/TTS/speaker_encoder/configs/config.json`

However I already had to modified several part of the code because it does not work, so I guess I am doing something not properly.
In the doc I can not find anything on creating a path/to/speaker.json, which is what I need I guess.

So my question are 
i) Can I generate sound directly with a command similar to the one I use
ii) How to generate the path/to/speaker.json

Best!

**Reply by erogol** [reply]
Voice cloning has no open endpoint implemented yet. You need to write your own code, unfortunately.

**Reply by hasankanaan26** [reply]
Hi guys, is voice cloning available as we speak?

**Reply by p0p4k** [reply]
@hasankanaan26 Do you mean by "as we speak" - "as real-time cloning" or "right now in this repo"?

**Reply by hasankanaan26** [reply]
Hi, I mean right now in this repo

**Reply by p0p4k** [reply]
Refer https://github.com/Edresson/YourTTS
That would be a good starting point for trying out the model, from there navigate to training your own model if you want to. Don't be afraid to ask here if you have any issues. Thanks.

**Reply by Edresson** [reply]
@hasankanaan26  The PR #1583 solve some bug in the voice conversion inference. After this PR is merged you will be able to use the released YourTTS model for voice conversion using the following command: 

```
tts --model_name tts_models/multilingual/multi-dataset/your_tts  --speaker_wav target_speaker_wav.wav --reference_wav  target_content_wav.wav --language_idx "en"
```
Considering the "target_content_wav.wav"  as the reference wave file to convert into the voice of the  "target_speaker_wav.wav" speaker.

**Reply by davidgarciaMontreal** [reply]
@Edresson I'd like to clone a voice from a Spanish speaker. I see the YourTTS model only support 3 languages for the language_idx. Is there a way to use this model for my purpose?
thanks
